ballerina/auth_desugar.bal

```diff
@@ -0,0 +1,162 @@
+        if scopes is string|string[] {
+            PermissionDeniedError? authz = handler.authorize(authn, scopes);
+            if authz is PermissionDeniedError {
+                return authz;
+            }
+            return true;
+        }
+        return true;
+    }
+}
+
+isolated function authenticateWithLdapUserStore(LdapUserStoreConfigWithScopes config, map<string|string[]> headers)
+ returns PermissionDeniedError|boolean? {
+    ListenerLdapUserStoreBasicAuthHandler handler = new(config.ldapUserStoreConfig);
+    auth:UserDetails|UnauthenticatedError authn = handler->authenticate(headers);
+    string|string[]? scopes = config?.scopes;
+    if authn is auth:UserDetails {
+        if scopes is string|string[] {
+            PermissionDeniedError? authz = handler->authorize(authn, scopes);
+            if authz is PermissionDeniedError {
+                return authz;
+            }
+            return true;
+        }
+        return true;
+    }
+}
+
+isolated function authenticateWithJwtValidator(JwtValidatorConfigWithScopes config, map<string|string[]> headers)
+ returns PermissionDeniedError|boolean? {
+    ListenerJwtAuthHandler handler = new(config.jwtValidatorConfig);
+    jwt:Payload|UnauthenticatedError authn = handler.authenticate(headers);
+    string|string[]? scopes = config?.scopes;
+    if authn is jwt:Payload {
+        if scopes is string|string[] {
+            PermissionDeniedError? authz = handler.authorize(authn, scopes);
+            if authz is PermissionDeniedError {
+                return authz;
+            }
+            return true;
+        }
+        return true;
+    }
+}
+
+isolated function authenticateWithOAuth2IntrospectionConfig(OAuth2IntrospectionConfigWithScopes config, map<string|string[]> headers)
+ returns PermissionDeniedError|boolean? {
+    // Here, config is OAuth2IntrospectionConfigWithScopes
+    ListenerOAuth2Handler handler = new(config.oauth2IntrospectionConfig);
+    oauth2:IntrospectionResponse|UnauthenticatedError|PermissionDeniedError auth = handler->authorize(headers, config?.scopes);
+    if auth is oauth2:IntrospectionResponse {
+        return true;
+    } else if auth is PermissionDeniedError {
+        return auth;
+    }
+}
+
+isolated function getListenerAuthConfig(Service serviceRef)
+                                        returns ListenerAuthConfig[]? {
+    ListenerAuthConfig[]? serviceAuthConfig = getServiceAuthConfig(serviceRef);
+    if serviceAuthConfig is ListenerAuthConfig[] {
+        return serviceAuthConfig;
+    }
+}
+
+isolated function getServiceAuthConfig(Service serviceRef) returns ListenerAuthConfig[]? {
+    typedesc<any> serviceTypeDesc = typeof serviceRef;
+    var serviceAnnotation = serviceTypeDesc.@AuthConfig;
+    if serviceAnnotation is () {
+        return;
+    }
+    AuthServiceConfig serviceConfig = <AuthServiceConfig>serviceAnnotation;
+    return serviceConfig?.auth;
+}
+
+isolated function sendError(Error authError) {
```

Comment:
How do we handle the 401/403 error response separately in gRPC domain with a single error method?

Reply:
We send PermissionDeniedError and UnauthenticatedError here

Reply:
Ack. Does the returning error of `externGetAuthorizationHeader` function also get converted to `UnauthenticatedError`?

Reply:
I send an `Internal` error [here](https://github.com/ballerina-platform/module-ballerina-grpc/blob/2b5be7c79b1b2e947c5351acc96d25881ef65b10/native/src/main/java/io/ballerina/stdlib/grpc/nativeimpl/caller/FunctionUtils.java#L196) with `Authorization header does not exist` message.